rate-limit events record in kubelet

jiangyaoguo opened the pull request:
Fixes #11912
1. Add rate limit for event record in pkg/client/unversioned/record.
2. Add parameter of event record qps and burst in kubelet.

jiangyaoguo commented:
@lavalamp @krousey

lavalamp commented on cmd/kubelet/app/server.go:
```diff
@@ -218,6 +220,8 @@ func (s *KubeletServer) AddFlags(fs *pflag.FlagSet) {
 	fs.StringVar(&s.HostNetworkSources, "host-network-sources", s.HostNetworkSources, "Comma-separated list of sources from which the Kubelet allows pods to use of host network. For all sources use \"*\" [default=\"file\"]")
 	fs.Float64Var(&s.RegistryPullQPS, "registry-qps", s.RegistryPullQPS, "If > 0, limit registry pull QPS to this value.  If 0, unlimited. [default=0.0]")
 	fs.IntVar(&s.RegistryBurst, "registry-burst", s.RegistryBurst, "Maximum size of a bursty pulls, temporarily allows pulls to burst to this number, while still not exceeding registry-qps.  Only used if --registry-qps > 0")
+	fs.Float32Var(&s.EventRecordQPS, "event-qps", s.EventRecordQPS, "If > 0, limit evnet record QPS to this value.  If 0, unlimited. [default=0.0]")
```
Change help string to: "If > 0, limit event creations per second to this value. If 0, unlimited. [default=0.0]"

lavalamp commented on cmd/kubelet/app/server.go:
```diff
@@ -218,6 +220,8 @@ func (s *KubeletServer) AddFlags(fs *pflag.FlagSet) {
 	fs.Float64Var(&s.RegistryPullQPS, "registry-qps", s.RegistryPullQPS, "If > 0, limit registry pull QPS to this value.  If 0, unlimited. [default=0.0]")
 	fs.IntVar(&s.RegistryBurst, "registry-burst", s.RegistryBurst, "Maximum size of a bursty pulls, temporarily allows pulls to burst to this number, while still not exceeding registry-qps.  Only used if --registry-qps > 0")
+	fs.Float32Var(&s.EventRecordQPS, "event-qps", s.EventRecordQPS, "If > 0, limit evnet record QPS to this value.  If 0, unlimited. [default=0.0]")
+	fs.IntVar(&s.EventBurst, "event-burst", s.EventBurst, "Maximum size of a bursty event records, temporarily allows event records to burst to this number, while still not exceeding event-qps.  Only used if --event-qps > 0")
```
"Temporarily allows event record creations to burst to this number, while still not exceeding event-qps.  Only used if --event-qps > 0"

lavalamp commented on cmd/kubelet/app/server.go:
```diff
@@ -638,7 +644,12 @@ func RunKubelet(kcfg *KubeletConfig, builder KubeletBuilder) error {
 		glog.V(4).Infof("Sending events to api server.")
-		eventBroadcaster.StartRecordingToSink(kcfg.KubeClient.Events(""))
+		if kcfg.EventRecordQPS == 0.0 {
+			eventBroadcaster.StartRecordingToSink(kcfg.KubeClient.Events(""), util.NewTokenBucketRateLimiter(5, 10))
```
Your help comment said that zero meant it'd be unlimited?

jiangyaoguo replied:
Sorry for commit with my test codes. It should be nil and unlimited.

lavalamp commented on cmd/kubelet/app/server.go:
```diff
@@ -638,7 +644,12 @@ func RunKubelet(kcfg *KubeletConfig, builder KubeletBuilder) error {
+		if kcfg.EventRecordQPS == 0.0 {
+			eventBroadcaster.StartRecordingToSink(kcfg.KubeClient.Events(""), util.NewTokenBucketRateLimiter(5, 10))
+		} else {
+			eventBroadcaster.StartRecordingToSink(kcfg.KubeClient.Events(""), util.NewTokenBucketRateLimiter(kcfg.EventRecordQPS, kcfg.EventBurst))
```
And what if it's negative?

lavalamp commented on pkg/client/unversioned/record/event.go:
```diff
@@ -222,6 +222,9 @@ func (eventBroadcaster *eventBroadcasterImpl) StartEventWatcher(eventHandler fun
 				continue
 			}
+			if ratelimiter != nil {
+				ratelimiter.Accept()
```
Will this cause those creating events to block? Under what circumstances?

jiangyaoguo replied:
@lavalamp Maybe I don't get your point. Accept() will wait for a  token generated by token bucket algorithm.

lavalamp replied:
Events are often created in the middle of time-sensitive loops. I'm worried about blocking enough events here that we back up the queue and start blocking the places that create events.

jiangyaoguo replied:
We have a default event queue with the length of 1000 and implicit max event limit is 5 QPS with no extra rate limiter. It works normally now. Suppose 1000 length of queue and 5 QPS can handle heavy traffic for at most 30s, when rate limit is 1QPS, the performance shoule be the same with about 1000+(5QPS-1QPS)＊30S length of queue(need more 120 length of queue).   But I don't think we need to increase length of queue.

jiangyaoguo replied:
@lavalamp

lavalamp commented:
We have default rate-limiting in the client. Should document how that interacts with this event QPS limit.

Like, I think right now if your client has a limit of 5 QPS and your events have a limit of 10 QPS, then effectively events are limited to 5 QPS and everything else could be starved. I'm not sure if this is a feature or a bug.

jiangyaoguo commented:
@lavalamp Thanks a lot for your review and advise. I made changes as you said and I'm not sure about where to document interacts of event QPS and client QPS. Should we suggest event QPS lower than client's default limit of 5 QPS?

lavalamp commented on cmd/kubelet/app/server.go:
```diff
@@ -638,7 +646,12 @@ func RunKubelet(kcfg *KubeletConfig, builder KubeletBuilder) error {
+		if kcfg.EventRecordQPS == 0.0 {
+			eventBroadcaster.StartRecordingToSink(kcfg.KubeClient.Events(""), nil)
+		} else {
+			eventBroadcaster.StartRecordingToSink(kcfg.KubeClient.Events(""), util.NewTokenBucketRateLimiter(kcfg.EventRecordQPS, kcfg.EventBurst))
```
Terribly sorry that I've been sitting on this, I guess I didn't see that you pushed changes. I have been thinking about this for a bit, and it occurs to me that there's another way to do this.

The solution you have here will work, but there's two downsides:
* People pass `nil` everywhere now.
* It's unpredictable how much quota will be left for non-event client usage.

An idea that fixes both of these is:
1. Make a copy of kcfg.KubeClient
2. Modify the copy's internal RESTClient.Throttle object (I think enough is public that you can do this)
3. Start recording to this copied client.

Then you don't have to add throttling to the event recorder at all, and the event QPS won't count against the regular QPS.

I'm sorry for not thinking of this to begin with. What do you think?

jiangyaoguo replied:
@lavalamp Thanks a lot for your idea. Totally agree with you. I think this will make event recorder behavior more clear. Will update this pr soon.

jiangyaoguo committed:
Rate limit events in kubelet

1. Add EvnetRecordQps and EventBurst parameter in kubelet.
2. If  EvnetRecordQps and EventBurst was set, rate limit events in kubelet
with a independent ratelimiter as setted.

lavalamp commented on hack/verify-flags/known-flags.txt:
```diff
@@ -1,269 +1,273 @@
-retry_time
-file_content_in_loop
-cpu-cfs-quota
+accept-hosts
```
Hm, I'm puzzled about why github shows such a bad diff on this file. Weird.

lavalamp commented:
This looks great, thank you very much for bearing with me!

LGTM
